# Pandas

package yang dibangun diatas Numpy dan bisa membuat grafik


In [1]:
import numpy as np
import pandas as pd

In [2]:
# data 1D - Series

arr = np.array([1, 4, 7, 11])

series1 = pd.Series(arr)

print("Numpy 1D berbentuk mirip List")
print(arr)

print("Pandas Series juga 1D tapi tampilannya berbeda")
print(series1)
print(type(series1))

Numpy 1D berbentuk mirip List
[ 1  4  7 11]
Pandas Series juga 1D tapi tampilannya berbeda
0     1
1     4
2     7
3    11
dtype: int32
<class 'pandas.core.series.Series'>


tipe data yang digunakan di dalam sebuah pandas datatype (Series maupun DataFrame) bisa numerik, bisa text, bisa date, dan banyak lainnya

Pandas lebih mirip excel dibandingkan Numpy array


In [3]:
# data 2D - DataFrame

arr = np.array([[4354, 6788, 234, 3460], [1234, 67568, 465, 7891]])

df1 = pd.DataFrame(arr)

print("Numpy 2D berbentuk Matrix")
print(arr)

print("Pandas DataFrame juga 2D tapi tampilannya lebih seperti spreadsheet")
print(df1)
print(type(df1))

Numpy 2D berbentuk Matrix
[[ 4354  6788   234  3460]
 [ 1234 67568   465  7891]]
Pandas DataFrame juga 2D tapi tampilannya lebih seperti spreadsheet
      0      1    2     3
0  4354   6788  234  3460
1  1234  67568  465  7891
<class 'pandas.core.frame.DataFrame'>


index rows dan index kolom bisa kita ganti juga jadi text, sehingga semakin mirip spreadsheet


In [4]:
df1.columns = ["A", "B", "C", "D"]

print(df1)

      A      B    C     D
0  4354   6788  234  3460
1  1234  67568  465  7891


In [5]:
df1.index = ["b1", "b2"]

print(df1)

       A      B    C     D
b1  4354   6788  234  3460
b2  1234  67568  465  7891


Pandas bisa memberitahu kita tipe data per kolom datanya. kalau di Series lsg 1 tipe data, kalau di DataFrame dia kasi taunya per kolom.


In [6]:
df1

A      B    C     D
b1  4354   6788  234  3460
b2  1234  67568  465  7891

In [7]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2 entries, b1 to b2
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   A       2 non-null      int32
 1   B       2 non-null      int32
 2   C       2 non-null      int32
 3   D       2 non-null      int32
dtypes: int32(4)
memory usage: 48.0+ bytes


In [8]:
df2 = pd.DataFrame([[1.0, 1, "a"], [2.0, 2, "b"], [3.0, 3, "c"]])

df2

0  1  2
0  1.0  1  a
1  2.0  2  b
2  3.0  3  c

In [9]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3 entries, 0 to 2
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   0       3 non-null      float64
 1   1       3 non-null      int64  
 2   2       3 non-null      object 
dtypes: float64(1), int64(1), object(1)
memory usage: 204.0+ bytes


Selain dengan list atau numpy array 2 dimensi, kita bisa buat DataFrame menggunakan dictionary juga


In [10]:
df3 = pd.DataFrame({
    "kolom1": ["data1", "data2", "data3"],
    "kolom2": [1, 2, 3]
})

df3

kolom1  kolom2
0  data1       1
1  data2       2
2  data3       3

In [11]:
df3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3 entries, 0 to 2
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   kolom1  3 non-null      object
 1   kolom2  3 non-null      int64 
dtypes: int64(1), object(1)
memory usage: 180.0+ bytes


## Data Loading

Kita bisa import data lokal maupun menggunakan link (online)

Contohnya, kita bisa download file csv dari Kaggle https://www.kaggle.com/datasets/thedevastator/mcdonalds-ice-cream-machines-broken-timeseries

dan kita import ke dalam notebook kita secara lokal


In [12]:
df_mcd = pd.read_csv("dataset\mcdonalds_dataset.csv")

df_mcd

lat        lon  alt  is_broken  is_active       dot state  \
0     -73.988281  40.718830    0      False       True   working    NY   
1     -74.005090  40.728794    0      False       True   working    NY   
2     -73.993408  40.729197    0      False       True   working    NY   
3     -73.985855  40.726555    0      False       True   working    NY   
4     -73.991692  40.691383    0       True       True    broken    NY   
...          ...        ...  ...        ...        ...       ...   ...   
16666  13.475643  52.514265    0      False      False  inactive   NaN   
16667  13.429812  54.076239    0      False      False  inactive   NaN   
16668   8.787059  53.100934    0      False      False  inactive   NaN   
16669  11.409059  53.628227    0      False      False  inactive   NaN   
16670  11.405999  53.903701    0      False      False  inactive   NaN   

             city                 street country             last_checked  
0        New York        114 Delancey St     USA  Checked 142 minutes ago  
1        New York          208 Varick St     USA  Checked 142 minutes ago  
2        New York           724 Broadway     USA  Checked 142 minutes ago  
3        New York            102 1st Ave     USA  Checked 142 minutes ago  
4        Brooklyn            82 Court St     USA  Checked 142 minutes ago  
...           ...                    ...     ...                      ...  
16666      Berlin  Frankfurter Allee 117      DE   Checked 31 minutes ago  
16667  Greifswald    Anklamer Landstr. 1      DE   Checked 31 minutes ago  
16668      Bremen    Waller Heerstr. 101      DE   Checked 31 minutes ago  
16669    Schwerin        Marienplatz 5-7      DE   Checked 31 minutes ago  
16670      Wismar    Zierower Landstr. 3      DE   Checked 31 minutes ago  

[16671 rows x 11 columns]

contoh file online (direct link ke file) https://github.com/stefkwan/dataset/raw/refs/heads/main/Sample%20-%20EU%20Superstore.xls


In [13]:
df_superstore = pd.read_excel(
    "https://github.com/stefkwan/dataset/raw/refs/heads/main/Sample%20-%20EU%20Superstore.xls")

df_superstore

Row ID         Order ID Order Date Dispatch Date   Delivery Mode  \
0          1  ES-2022-1311038 2022-02-07    2022-02-11  Standard Class   
1          2  ES-2022-1311038 2022-02-07    2022-02-11  Standard Class   
2          3  ES-2022-1311038 2022-02-07    2022-02-11  Standard Class   
3          4  ES-2022-1311038 2022-02-07    2022-02-11  Standard Class   
4          5  ES-2022-1311038 2022-02-07    2022-02-11  Standard Class   
...      ...              ...        ...           ...             ...   
9995    9996  ES-2019-3780358 2019-11-16    2019-11-19     First Class   
9996    9997  ES-2020-5371207 2020-01-25    2020-01-29  Standard Class   
9997    9998  ES-2020-5371207 2020-01-25    2020-01-29  Standard Class   
9998    9999  ES-2020-5371207 2020-01-25    2020-01-29  Standard Class   
9999   10000  ES-2020-5371207 2020-01-25    2020-01-29  Standard Class   

     Customer ID      Customer Name      Segment     City State/Province  \
0       AS-10045     Aaron Smayling    Corporate    Leeds        England   
1       AS-10045     Aaron Smayling    Corporate    Leeds        England   
2       AS-10045     Aaron Smayling    Corporate    Leeds        England   
3       AS-10045     Aaron Smayling    Corporate    Leeds        England   
4       AS-10045     Aaron Smayling    Corporate    Leeds        England   
...          ...                ...          ...      ...            ...   
9995    CC-12100    Chad Cunningham  Home Office   Rovigo         Veneto   
9996    GB-14575  Giulietta Baptist     Consumer  Hamburg        Hamburg   
9997    GB-14575  Giulietta Baptist     Consumer  Hamburg        Hamburg   
9998    GB-14575  Giulietta Baptist     Consumer  Hamburg        Hamburg   
9999    GB-14575  Giulietta Baptist     Consumer  Hamburg        Hamburg   

      Country/Region   Region       Product ID         Category Sub-Category  \
0     United Kingdom    North  OFF-ST-10000988  Office Supplies      Storage   
1     United Kingdom    North  TEC-AC-10004144       Technology  Accessories   
2     United Kingdom    North  OFF-LA-10001915  Office Supplies       Labels   
3     United Kingdom    North  OFF-ST-10004550  Office Supplies      Storage   
4     United Kingdom    North  TEC-AC-10004068       Technology  Accessories   
...              ...      ...              ...              ...          ...   
9995           Italy    South  OFF-BI-10000346  Office Supplies      Binders   
9996         Germany  Central  OFF-BI-10002083  Office Supplies      Binders   
9997         Germany  Central  OFF-SU-10003211  Office Supplies     Supplies   
9998         Germany  Central  TEC-PH-10001664       Technology       Phones   
9999         Germany  Central  OFF-AR-10001228  Office Supplies          Art   

                                    Product Name   Sales  Quantity  Discount  \
0                         Fellowes Folders, Blue   79.20         3       0.0   
1              SanDisk Numeric Keypad, Bluetooth  388.92         7       0.0   
2     Avery Legal Exhibit Labels, 5000 Label Set   35.19         3       0.0   
3                   Fellowes Folders, Wire Frame   50.94         2       0.0   
4                       Memorex Memory Card, USB  307.44         3       0.0   
...                                          ...     ...       ...       ...   
9995                   Ibico 3-Hole Punch, Clear   59.82         2       0.0   
9996           Acco Hole Reinforcements, Economy    6.66         1       0.0   
9997                           Acme Ruler, Steel   96.66         6       0.0   
9998             Motorola Office Telephone, VoIP  502.95         7       0.0   
9999                Stanley Markers, Water Color  152.28         6       0.0   

      Profit  
0      39.60  
1       0.00  
2      16.11  
3      13.20  
4      73.71  
...      ...  
9995   23.88  
9996    0.57  
9997   28.98  
9998  221.13  
9999   50.22  

[10000 rows x 20 columns]

In [14]:
# pakai .info() untuk cek bentuk data

df_mcd.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16671 entries, 0 to 16670
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   lat           16671 non-null  float64
 1   lon           16671 non-null  float64
 2   alt           16671 non-null  int64  
 3   is_broken     16671 non-null  bool   
 4   is_active     16671 non-null  bool   
 5   dot           16671 non-null  object 
 6   state         12725 non-null  object 
 7   city          16663 non-null  object 
 8   street        16671 non-null  object 
 9   country       16671 non-null  object 
 10  last_checked  16671 non-null  object 
dtypes: bool(2), float64(2), int64(1), object(6)
memory usage: 1.2+ MB


In [15]:
df_superstore.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 20 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   Row ID          10000 non-null  int64         
 1   Order ID        10000 non-null  object        
 2   Order Date      10000 non-null  datetime64[ns]
 3   Dispatch Date   10000 non-null  datetime64[ns]
 4   Delivery Mode   10000 non-null  object        
 5   Customer ID     10000 non-null  object        
 6   Customer Name   10000 non-null  object        
 7   Segment         10000 non-null  object        
 8   City            10000 non-null  object        
 9   State/Province  10000 non-null  object        
 10  Country/Region  10000 non-null  object        
 11  Region          10000 non-null  object        
 12  Product ID      10000 non-null  object        
 13  Category        10000 non-null  object        
 14  Sub-Category    10000 non-null  object        
 15  Pro

## Statistics in Pandas

Kita bisa gunakan fungsi bawaan pandas untuk hitung agregasi data

contohnya, rata2 sales

selain itu ada "groupby" untuk mengelompokkan data sebelum di agregasi

contohnya, rata2 sales per segment


In [16]:
# rata2 sales

df_superstore['Sales'].mean()

293.80890615

In [17]:
df_superstore.agg({"Sales": "mean"})

Sales    293.808906
dtype: float64

In [18]:
df_superstore.groupby("Segment").agg({"Sales": "mean"})

Sales
Segment                
Consumer     294.970350
Corporate    298.995216
Home Office  281.154027

bahkan untuk beberapa grup dan beberapa agregasi juga bisa

misalnya, per Segment dan per Region, lalu mau lihat "sum" dan "mean"


In [20]:
df_superstore.groupby(["Segment", "Region"]).agg(
    {"Sales": ["mean", "sum"]})

Sales             
                           mean          sum
Segment     Region                          
Consumer    Central  298.064136  911182.0635
            North    296.198089  322855.9170
            South    284.579651  295678.2570
Corporate   Central  295.898129  516934.0305
            North    291.265596  195439.2150
            South    315.075924  207635.0340
Home Office Central  287.265741  292436.5245
            North    282.315403  107279.8530
            South    261.499018   88648.1670

In [23]:
# semua kolom, khusus kolom numerik
df_superstore[["Sales", "Quantity", "Discount", "Profit", "Segment", "Region"]
              ].groupby(["Segment", "Region"]).agg(["mean", "sum"])

Sales               Quantity         Discount  \
                           mean          sum      mean    sum      mean   
Segment     Region                                                        
Consumer    Central  298.064136  911182.0635  3.806673  11637  0.095976   
            North    296.198089  322855.9170  3.700917   4034  0.145413   
            South    284.579651  295678.2570  3.724735   3870  0.100192   
Corporate   Central  295.898129  516934.0305  3.782484   6608  0.084602   
            North    291.265596  195439.2150  3.821162   2564  0.135171   
            South    315.075924  207635.0340  3.737481   2463  0.102580   
Home Office Central  287.265741  292436.5245  3.797642   3866  0.088949   
            North    282.315403  107279.8530  3.715789   1412  0.125000   
            South    261.499018   88648.1670  3.890855   1319  0.091150   

                                Profit               
                        sum       mean          sum  
Segment     Region                                   
Consumer    Central  293.40  36.391643  111249.2535  
            North    158.50  41.502355   45237.5670  
            South    104.10  30.992192   32200.8870  
Corporate   Central  147.80  39.603818   69187.8705  
            North     90.70  49.714940   33358.7250  
            South     67.60  31.634877   20847.3840  
Home Office Central   90.55  34.476370   35096.9445  
            North     47.50  34.693587   13183.5630  
            South     30.90  36.777425   12467.5470

## Slicing dan Indexing

dari sebuah DataFrame kita bisa ambil sebagian datanya, misalnya kolom2 tertentu, atau baris2 tertentu, atau kolom dan baris tertentu saja

- kita bisa lsg minta kolom tertentu dengan menggunakan judul kolom.
- untuk minta baris tertentu, kita perlu tambahkan ".loc" atau ".iloc"


In [24]:
# contohnya, kita ambil data superstore khusus kolom kolom tertentu
# "Sales", "Quantity", "Discount", "Profit", "Segment", "Region"

df_superstore[["Sales", "Quantity", "Discount", "Profit", "Segment", "Region"]]

Sales  Quantity  Discount  Profit      Segment   Region
0      79.20         3       0.0   39.60    Corporate    North
1     388.92         7       0.0    0.00    Corporate    North
2      35.19         3       0.0   16.11    Corporate    North
3      50.94         2       0.0   13.20    Corporate    North
4     307.44         3       0.0   73.71    Corporate    North
...      ...       ...       ...     ...          ...      ...
9995   59.82         2       0.0   23.88  Home Office    South
9996    6.66         1       0.0    0.57     Consumer  Central
9997   96.66         6       0.0   28.98     Consumer  Central
9998  502.95         7       0.0  221.13     Consumer  Central
9999  152.28         6       0.0   50.22     Consumer  Central

[10000 rows x 6 columns]

In [25]:
# untuk ambil data 1 kolom saja
df_superstore[["Sales"]]

Sales
0      79.20
1     388.92
2      35.19
3      50.94
4     307.44
...      ...
9995   59.82
9996    6.66
9997   96.66
9998  502.95
9999  152.28

[10000 rows x 1 columns]

In [ ]:
# untuk ambil data 1 kolom saja, kalau tidak menggunakan list, tipe data berubah jadi Series
df_superstore["Sales"]

0        79.20
1       388.92
2        35.19
3        50.94
4       307.44
         ...  
9995     59.82
9996      6.66
9997     96.66
9998    502.95
9999    152.28
Name: Sales, Length: 10000, dtype: float64

In [ ]:
# ambil 1 baris saja, dan tipe data menjadi Series
df_superstore.loc[0]

Row ID                                 1
Order ID                 ES-2022-1311038
Order Date           2022-02-07 00:00:00
Dispatch Date        2022-02-11 00:00:00
Delivery Mode             Standard Class
Customer ID                     AS-10045
Customer Name             Aaron Smayling
Segment                        Corporate
City                               Leeds
State/Province                   England
Country/Region            United Kingdom
Region                             North
Product ID               OFF-ST-10000988
Category                 Office Supplies
Sub-Category                     Storage
Product Name      Fellowes Folders, Blue
Sales                               79.2
Quantity                               3
Discount                             0.0
Profit                              39.6
Name: 0, dtype: object

In [32]:
# hasil ambil data tetap DataFrame
df_superstore.loc[[0]]

Row ID         Order ID Order Date Dispatch Date   Delivery Mode  \
0       1  ES-2022-1311038 2022-02-07    2022-02-11  Standard Class   

  Customer ID   Customer Name    Segment   City State/Province  \
0    AS-10045  Aaron Smayling  Corporate  Leeds        England   

   Country/Region Region       Product ID         Category Sub-Category  \
0  United Kingdom  North  OFF-ST-10000988  Office Supplies      Storage   

             Product Name  Sales  Quantity  Discount  Profit  
0  Fellowes Folders, Blue   79.2         3       0.0    39.6

In [33]:
# ambil beberapa baris
df_superstore.loc[[0, 1, 2]]

Row ID         Order ID Order Date Dispatch Date   Delivery Mode  \
0       1  ES-2022-1311038 2022-02-07    2022-02-11  Standard Class   
1       2  ES-2022-1311038 2022-02-07    2022-02-11  Standard Class   
2       3  ES-2022-1311038 2022-02-07    2022-02-11  Standard Class   

  Customer ID   Customer Name    Segment   City State/Province  \
0    AS-10045  Aaron Smayling  Corporate  Leeds        England   
1    AS-10045  Aaron Smayling  Corporate  Leeds        England   
2    AS-10045  Aaron Smayling  Corporate  Leeds        England   

   Country/Region Region       Product ID         Category Sub-Category  \
0  United Kingdom  North  OFF-ST-10000988  Office Supplies      Storage   
1  United Kingdom  North  TEC-AC-10004144       Technology  Accessories   
2  United Kingdom  North  OFF-LA-10001915  Office Supplies       Labels   

                                 Product Name   Sales  Quantity  Discount  \
0                      Fellowes Folders, Blue   79.20         3       0.0   
1           SanDisk Numeric Keypad, Bluetooth  388.92         7       0.0   
2  Avery Legal Exhibit Labels, 5000 Label Set   35.19         3       0.0   

   Profit  
0   39.60  
1    0.00  
2   16.11

karena kita akan sering "intip" bentuk data dengan melihat beberapa baris pertama dan beberapa baris terakhir, ada built-in function

- .head(10)
- .tail(3)

karena function, jadi kita pakai ()


In [34]:
df_superstore.head(3)

Row ID         Order ID Order Date Dispatch Date   Delivery Mode  \
0       1  ES-2022-1311038 2022-02-07    2022-02-11  Standard Class   
1       2  ES-2022-1311038 2022-02-07    2022-02-11  Standard Class   
2       3  ES-2022-1311038 2022-02-07    2022-02-11  Standard Class   

  Customer ID   Customer Name    Segment   City State/Province  \
0    AS-10045  Aaron Smayling  Corporate  Leeds        England   
1    AS-10045  Aaron Smayling  Corporate  Leeds        England   
2    AS-10045  Aaron Smayling  Corporate  Leeds        England   

   Country/Region Region       Product ID         Category Sub-Category  \
0  United Kingdom  North  OFF-ST-10000988  Office Supplies      Storage   
1  United Kingdom  North  TEC-AC-10004144       Technology  Accessories   
2  United Kingdom  North  OFF-LA-10001915  Office Supplies       Labels   

                                 Product Name   Sales  Quantity  Discount  \
0                      Fellowes Folders, Blue   79.20         3       0.0   
1           SanDisk Numeric Keypad, Bluetooth  388.92         7       0.0   
2  Avery Legal Exhibit Labels, 5000 Label Set   35.19         3       0.0   

   Profit  
0   39.60  
1    0.00  
2   16.11

In [ ]:
df_superstore.tail()

# kalau tidak diisi mau berapa baris, dia default ambil 5 baris

Row ID         Order ID Order Date Dispatch Date   Delivery Mode  \
9995    9996  ES-2019-3780358 2019-11-16    2019-11-19     First Class   
9996    9997  ES-2020-5371207 2020-01-25    2020-01-29  Standard Class   
9997    9998  ES-2020-5371207 2020-01-25    2020-01-29  Standard Class   
9998    9999  ES-2020-5371207 2020-01-25    2020-01-29  Standard Class   
9999   10000  ES-2020-5371207 2020-01-25    2020-01-29  Standard Class   

     Customer ID      Customer Name      Segment     City State/Province  \
9995    CC-12100    Chad Cunningham  Home Office   Rovigo         Veneto   
9996    GB-14575  Giulietta Baptist     Consumer  Hamburg        Hamburg   
9997    GB-14575  Giulietta Baptist     Consumer  Hamburg        Hamburg   
9998    GB-14575  Giulietta Baptist     Consumer  Hamburg        Hamburg   
9999    GB-14575  Giulietta Baptist     Consumer  Hamburg        Hamburg   

     Country/Region   Region       Product ID         Category Sub-Category  \
9995          Italy    South  OFF-BI-10000346  Office Supplies      Binders   
9996        Germany  Central  OFF-BI-10002083  Office Supplies      Binders   
9997        Germany  Central  OFF-SU-10003211  Office Supplies     Supplies   
9998        Germany  Central  TEC-PH-10001664       Technology       Phones   
9999        Germany  Central  OFF-AR-10001228  Office Supplies          Art   

                           Product Name   Sales  Quantity  Discount  Profit  
9995          Ibico 3-Hole Punch, Clear   59.82         2       0.0   23.88  
9996  Acco Hole Reinforcements, Economy    6.66         1       0.0    0.57  
9997                  Acme Ruler, Steel   96.66         6       0.0   28.98  
9998    Motorola Office Telephone, VoIP  502.95         7       0.0  221.13  
9999       Stanley Markers, Water Color  152.28         6       0.0   50.22

## Query

filter data dengan kondisi

misalnya kita mau ambil data khusus Customer Name "Chad Cunningham"


In [38]:
kondisi1 = df_superstore['Customer Name'] == "Chad Cunningham"

df_chad = df_superstore[kondisi1]
df_chad

Row ID         Order ID Order Date Dispatch Date   Delivery Mode  \
550      551  ES-2021-2934598 2021-07-12    2021-07-16  Standard Class   
551      552  ES-2021-2934598 2021-07-12    2021-07-16  Standard Class   
552      553  ES-2021-2934598 2021-07-12    2021-07-16  Standard Class   
553      554  ES-2021-2934598 2021-07-12    2021-07-16  Standard Class   
554      555  ES-2021-2934598 2021-07-12    2021-07-16  Standard Class   
555      556  ES-2021-2934598 2021-07-12    2021-07-16  Standard Class   
556      557  ES-2021-2934598 2021-07-12    2021-07-16  Standard Class   
557      558  ES-2021-2934598 2021-07-12    2021-07-16  Standard Class   
4124    4125  ES-2022-5554891 2022-06-27    2022-06-27        Same Day   
5210    5211  ES-2022-5746112 2022-08-06    2022-08-12  Standard Class   
5712    5713  ES-2021-4117792 2021-12-07    2021-12-10     First Class   
8315    1459  ES-2020-4984653 2020-01-06    2020-01-06        Same Day   
8423    1460  ES-2020-4984653 2020-01-06    2020-01-06        Same Day   
8479    1461  ES-2020-4984653 2020-01-06    2020-01-06        Same Day   
9293    9294  ES-2020-3534614 2020-02-10    2020-02-14  Standard Class   
9995    9996  ES-2019-3780358 2019-11-16    2019-11-19     First Class   

     Customer ID    Customer Name      Segment       City  \
550     CC-12100  Chad Cunningham  Home Office       Rome   
551     CC-12100  Chad Cunningham  Home Office       Rome   
552     CC-12100  Chad Cunningham  Home Office       Rome   
553     CC-12100  Chad Cunningham  Home Office       Rome   
554     CC-12100  Chad Cunningham  Home Office       Rome   
555     CC-12100  Chad Cunningham  Home Office       Rome   
556     CC-12100  Chad Cunningham  Home Office       Rome   
557     CC-12100  Chad Cunningham  Home Office       Rome   
4124    CC-12100  Chad Cunningham  Home Office   Grenoble   
5210    CC-12100  Chad Cunningham  Home Office  Stavanger   
5712    CC-12100  Chad Cunningham  Home Office      Paris   
8315    CC-12100  Chad Cunningham  Home Office      Rouen   
8423    CC-12100  Chad Cunningham  Home Office      Rouen   
8479    CC-12100  Chad Cunningham  Home Office      Rouen   
9293    CC-12100  Chad Cunningham  Home Office      Reims   
9995    CC-12100  Chad Cunningham  Home Office     Rovigo   

                         State/Province Country/Region   Region  \
550                               Lazio          Italy    South   
551                               Lazio          Italy    South   
552                               Lazio          Italy    South   
553                               Lazio          Italy    South   
554                               Lazio          Italy    South   
555                               Lazio          Italy    South   
556                               Lazio          Italy    South   
557                               Lazio          Italy    South   
4124               Auvergne-Rhône-Alpes         France  Central   
5210                           Rogaland         Norway    North   
5712                      Ile-de-France         France  Central   
8315                           Normandy         France  Central   
8423                           Normandy         France  Central   
8479                           Normandy         France  Central   
9293  Alsace-Champagne-Ardenne-Lorraine         France  Central   
9995                             Veneto          Italy    South   

           Product ID         Category Sub-Category  \
550   OFF-LA-10004744  Office Supplies       Labels   
551   TEC-CO-10002872       Technology      Copiers   
552   OFF-ST-10002566  Office Supplies      Storage   
553   OFF-AR-10001002  Office Supplies          Art   
554   OFF-SU-10003115  Office Supplies     Supplies   
555   OFF-AR-10000491  Office Supplies          Art   
556   TEC-CO-10004365       Technology      Copiers   
557   OFF-BI-10000329  Office Supplies      Binders   
4124  TEC-CO-10004567       Technology      Copiers   
5210  OFF-AR-100004

In [39]:
df_chad.info()

<class 'pandas.core.frame.DataFrame'>
Index: 16 entries, 550 to 9995
Data columns (total 20 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   Row ID          16 non-null     int64         
 1   Order ID        16 non-null     object        
 2   Order Date      16 non-null     datetime64[ns]
 3   Dispatch Date   16 non-null     datetime64[ns]
 4   Delivery Mode   16 non-null     object        
 5   Customer ID     16 non-null     object        
 6   Customer Name   16 non-null     object        
 7   Segment         16 non-null     object        
 8   City            16 non-null     object        
 9   State/Province  16 non-null     object        
 10  Country/Region  16 non-null     object        
 11  Region          16 non-null     object        
 12  Product ID      16 non-null     object        
 13  Category        16 non-null     object        
 14  Sub-Category    16 non-null     object        
 15  Product N

In [40]:
# double check data di kolom Customer Name
df_chad['Customer Name'].unique()

array(['Chad Cunningham'], dtype=object)

Untuk cari text yang _wildcard_ tidak exact match, kita bisa pakai ".str.contains()"

- .str.contains() # ada text ini
- .str.startswith() # dimulai dengan
- .str.endswith() # diakhiri dengan


In [42]:
# contoh: semua customer Chad

kondisi2 = df_superstore['Customer Name'].str.contains("Chad")

df_chad2 = df_superstore[kondisi2]

df_chad2['Customer Name'].unique()

array(['Chad Cunningham', 'Chad Sievert', 'Chad McGuire'], dtype=object)

untuk angka, query nya bisa juga untuk angka tertentu, lebih besar, lebih kecil, atau bukan angka tertentu


Kita juga bisa gunakan lebih dari 1 kondisi untuk filter nya

misalnya, selain Chad kita juga mau ambil semua Aaron


In [ ]:
kondisi2 = df_superstore['Customer Name'].str.contains("Chad")
kondisi3 = df_superstore['Customer Name'].str.contains("Aaron")

# | or, artinya salah satu kondisi terpenuhi
df_chadaaron = df_superstore[kondisi2 | kondisi3]

df_chadaaron['Customer Name'].unique()

array(['Aaron Smayling', 'Chad Cunningham', 'Aaron Bergman',
       'Aaron Hawkins', 'Chad Sievert', 'Chad McGuire'], dtype=object)

In [ ]:
# misalnya selain Chad, dia harus dari Segment "Home Office"
kondisi2 = df_superstore['Customer Name'].str.contains("Chad")
kondisi4 = df_superstore['Segment'] == "Home Office"

# & and, kedua kondisi harus terpenuhi
df_chadhome = df_superstore[kondisi2 & kondisi4]

print(df_chadhome['Customer Name'].unique())
print(df_chadhome['Segment'].unique())

['Chad Cunningham']
['Home Office']


## Explorasi Data

- df.head() # baris awal, 5 default, atau n bebas mau berapa baris
- df.tail() # baris akhir
- df.shape # dataframe ini punya berapa baris dan berapa kolom
- df.info() # selain jumlah baris dan kolom, juga list judul2 kolom, dan tipe data per kolom
- df.describe() # summary statistik dataframe kita


In [47]:
df_mcd.head()

lat        lon  alt  is_broken  is_active      dot state      city  \
0 -73.988281  40.718830    0      False       True  working    NY  New York   
1 -74.005090  40.728794    0      False       True  working    NY  New York   
2 -73.993408  40.729197    0      False       True  working    NY  New York   
3 -73.985855  40.726555    0      False       True  working    NY  New York   
4 -73.991692  40.691383    0       True       True   broken    NY  Brooklyn   

            street country             last_checked  
0  114 Delancey St     USA  Checked 142 minutes ago  
1    208 Varick St     USA  Checked 142 minutes ago  
2     724 Broadway     USA  Checked 142 minutes ago  
3      102 1st Ave     USA  Checked 142 minutes ago  
4      82 Court St     USA  Checked 142 minutes ago

In [48]:
df_mcd.head(1)

lat       lon  alt  is_broken  is_active      dot state      city  \
0 -73.988281  40.71883    0      False       True  working    NY  New York   

            street country             last_checked  
0  114 Delancey St     USA  Checked 142 minutes ago

In [49]:
df_mcd.tail(3)

lat        lon  alt  is_broken  is_active       dot state  \
16668   8.787059  53.100934    0      False      False  inactive   NaN   
16669  11.409059  53.628227    0      False      False  inactive   NaN   
16670  11.405999  53.903701    0      False      False  inactive   NaN   

           city               street country            last_checked  
16668    Bremen  Waller Heerstr. 101      DE  Checked 31 minutes ago  
16669  Schwerin      Marienplatz 5-7      DE  Checked 31 minutes ago  
16670    Wismar  Zierower Landstr. 3      DE  Checked 31 minutes ago

In [50]:
df_mcd.shape

(16671, 11)

In [51]:
df_mcd.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16671 entries, 0 to 16670
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   lat           16671 non-null  float64
 1   lon           16671 non-null  float64
 2   alt           16671 non-null  int64  
 3   is_broken     16671 non-null  bool   
 4   is_active     16671 non-null  bool   
 5   dot           16671 non-null  object 
 6   state         12725 non-null  object 
 7   city          16663 non-null  object 
 8   street        16671 non-null  object 
 9   country       16671 non-null  object 
 10  last_checked  16671 non-null  object 
dtypes: bool(2), float64(2), int64(1), object(6)
memory usage: 1.2+ MB


In [52]:
df_mcd.describe()

lat           lon      alt
count  16671.000000  16671.000000  16671.0
mean     -76.600800     40.406232      0.0
std       37.309030      7.250517      0.0
min     -159.368738     19.517590      0.0
25%      -95.572842     34.927834      0.0
50%      -83.780231     40.213555      0.0
75%      -75.261739     44.860442      0.0
max       14.968594     64.859406      0.0

In [53]:
df_superstore.describe()

Row ID                  Order Date        Dispatch Date  \
count  10000.00000                       10000                10000   
mean    5000.50000  2021-05-14 15:05:45.600000  2021-05-18 15:14:24   
min        1.00000         2019-01-01 00:00:00  2019-01-05 00:00:00   
25%     2500.75000         2020-06-30 00:00:00  2020-07-05 00:00:00   
50%     5000.50000         2021-07-09 00:00:00  2021-07-11 00:00:00   
75%     7500.25000         2022-06-03 00:00:00  2022-06-06 00:00:00   
max    10000.00000         2022-12-31 00:00:00  2023-01-05 00:00:00   
std     2886.89568                         NaN                  NaN   

              Sales      Quantity      Discount        Profit  
count  10000.000000  10000.000000  10000.000000  10000.000000  
mean     293.808906      3.777300      0.103105     37.282974  
min        2.955000      1.000000      0.000000  -3059.820000  
25%       49.462500      2.000000      0.000000      1.320000  
50%      119.355000      3.000000      0.000000     14.220000  
75%      320.708625      5.000000      0.100000     48.510000  
max     7958.580000     14.000000      0.850000   3979.080000  
std      486.055635      2.203268      0.174782    178.100384

## Explorasi Data (lanjut)

- df.head() # baris awal, 5 default, atau n bebas mau berapa baris
- df.tail() # baris akhir
- df.shape # dataframe ini punya berapa baris dan berapa kolom
- df.info() # selain jumlah baris dan kolom, juga list judul2 kolom, dan tipe data per kolom
- df.describe() # summary statistik dataframe kita
- df.columns # mendapat judul2 kolom
- df.index # mendapat judul2 baris
- df.dtypes # mendapat tipe2 data kolom
- df.isnull() # mengecek datanya ada yang hilang (missing values) atau tidak

- df.nunique() # di kolom ini, ada berapa unique values, ada berapa nilai yang sifatnya unik
- df.unique() # value2 unik nya apa saja
- df.value_counts() # masing2 value unik muncul berapa kali, count nya berapa


In [54]:
df_mcd.columns

Index(['lat', 'lon', 'alt', 'is_broken', 'is_active', 'dot', 'state', 'city',
       'street', 'country', 'last_checked'],
      dtype='object')

In [55]:
df_mcd.index

RangeIndex(start=0, stop=16671, step=1)

kalau judul barisnya adalah angka urut, dia akan muncul sebagai RangeIndex


In [58]:
df_mcd.dtypes

lat             float64
lon             float64
alt               int64
is_broken          bool
is_active          bool
dot              object
state            object
city             object
street           object
country          object
last_checked     object
dtype: object

In [61]:
df_mcd.isnull()

lat    lon    alt  is_broken  is_active    dot  state   city  street  \
0      False  False  False      False      False  False  False  False   False   
1      False  False  False      False      False  False  False  False   False   
2      False  False  False      False      False  False  False  False   False   
3      False  False  False      False      False  False  False  False   False   
4      False  False  False      False      False  False  False  False   False   
...      ...    ...    ...        ...        ...    ...    ...    ...     ...   
16666  False  False  False      False      False  False   True  False   False   
16667  False  False  False      False      False  False   True  False   False   
16668  False  False  False      False      False  False   True  False   False   
16669  False  False  False      False      False  False   True  False   False   
16670  False  False  False      False      False  False   True  False   False   

       country  last_checked  
0        False         False  
1        False         False  
2        False         False  
3        False         False  
4        False         False  
...        ...           ...  
16666    False         False  
16667    False         False  
16668    False         False  
16669    False         False  
16670    False         False  

[16671 rows x 11 columns]

In [62]:
# umumnya digabung dengan sum() untuk lihat total data yang hilang per kolom
df_mcd.isnull().sum()

lat                0
lon                0
alt                0
is_broken          0
is_active          0
dot                0
state           3946
city               8
street             0
country            0
last_checked       0
dtype: int64

In [63]:
df_superstore['Customer Name'].nunique()

795

ada 795 customer dengan nama yang unik


In [64]:
df_superstore['Customer ID'].nunique()

795

In [65]:
# masing2 customer muncul berapa kali
df_superstore['Customer Name'].value_counts()

Customer Name
Becky Castell    37
John Grady       35
Mathew Reese     34
Michael Paige    33
John Murray      32
                 ..
Anne Pryor        2
Amy Hunt          2
Ruben Dartt       1
Sam Zeldin        1
Chuck Clark       1
Name: count, Length: 795, dtype: int64

In [66]:
# contoh lain, Segment, ada berapa segment, dan segment nya apa saja
df_superstore['Segment'].nunique()

3

In [67]:
df_superstore['Segment'].unique()

array(['Corporate', 'Consumer', 'Home Office'], dtype=object)

In [68]:
df_superstore['Segment'].value_counts()

Segment
Consumer       5186
Corporate      3077
Home Office    1737
Name: count, dtype: int64

kita bisa tau kalau ada 3 Segment, yaitu 'Corporate', 'Consumer', dan 'Home Office', masing2 muncul 5186, 3077, dan 1737 kali dalam dataframe kita.


## Review slicing

ambil kolom tertentu dengan judul kolomnya:

- df["judul kolom"] # Series
- df[["judul kolom"]] # DataFrame, dan bisa ambil lebih dari 1 kolom

ambil baris tertentu:

- df.head() # baris awal
- df.tail() # baris akhir
- df.loc() # menggunakan judul baris (index)
- df.iloc() # menggunakan posisi baris (angka index mulai dari 0)


In [69]:
df1

A      B    C     D
b1  4354   6788  234  3460
b2  1234  67568  465  7891

In [ ]:
df1["A"]  # jadi Series

b1    4354
b2    1234
Name: A, dtype: int32

In [ ]:
df1[["A"]]  # tetap DataFrame

A
b1  4354
b2  1234

In [ ]:
df1[["A", "B"]]  # tetap DataFrame, lebih dari 1 kolom

A      B
b1  4354   6788
b2  1234  67568

In [ ]:
df1.loc["b1"]  # jadi Series

A    4354
B    6788
C     234
D    3460
Name: b1, dtype: int32

In [ ]:
df1.loc[["b1"]]  # tetap DataFrame

A     B    C     D
b1  4354  6788  234  3460

In [ ]:
df1.loc[["b1", "b2"]]  # tetap DataFrame, lebih dari 1 baris

A      B    C     D
b1  4354   6788  234  3460
b2  1234  67568  465  7891

In [77]:
df1.iloc[0]  # jadi Series

A    4354
B    6788
C     234
D    3460
Name: b1, dtype: int32

In [78]:
df1.iloc[[0]]  # tetap DataFrame

A     B    C     D
b1  4354  6788  234  3460

In [79]:
df1.iloc[[0, 1]]  # tetap DataFrame, lebih dari 1 baris

A      B    C     D
b1  4354   6788  234  3460
b2  1234  67568  465  7891

## Query (lagi)

kondisi sebenarnya adalah Series/DataFrame juga yang isinya boolean semua

kalau kita filter per data sebuah kolom, kita akan dapat baris2 tertentu saja, baris2 yang True


In [83]:
df_mcd['city'] == "Brooklyn"

0        False
1        False
2        False
3        False
4         True
         ...  
16666    False
16667    False
16668    False
16669    False
16670    False
Name: city, Length: 16671, dtype: bool

In [84]:
df_mcd[df_mcd['city'] == "Brooklyn"]

lat        lon  alt  is_broken  is_active       dot state  \
4    -73.991692  40.691383    0       True       True    broken    NY   
10   -73.980670  40.688963    0       True       True    broken    NY   
19   -73.959017  40.708930    0      False       True   working    NY   
23   -73.999161  40.673519    0      False       True   working    NY   
27   -73.954109  40.730389    0       True       True    broken    NY   
28   -73.966927  40.681351    0       True       True    broken    NY   
31   -73.986854  40.669625    0      False       True   working    NY   
35   -73.943154  40.711849    0       True       True    broken    NY   
39   -73.943489  40.701332    0      False       True   working    NY   
40   -73.940483  40.722080    0      False       True   working    NY   
44   -73.950340  40.680573    0       True       True    broken    NY   
48   -74.005402  40.653751    0      False       True   working    NY   
49   -73.940102  40.680096    0      False       True   working    NY   
50   -73.959984  40.663731    0       True       True    broken    NY   
54   -73.919655  40.699417    0      False       True   working    NY   
55   -73.961690  40.654762    0      False       True   working    NY   
58   -73.921887  40.689106    0      False       True   working    NY   
59   -73.987686  40.644569    0      False       True   working    NY   
60   -73.930779  40.669163    0      False       True   working    NY   
61   -73.922554  40.677324    0      False       True   working    NY   
67   -73.930542  40.656101    0      False       True   working    NY   
68   -74.027039  40.633682    0      False       True   working    NY   
70   -73.968124  40.635811    0      False       True   working    NY   
71   -73.912406  40.669307    0      False       True   working    NY   
76   -73.923203  40.652718    0      False       True   working    NY   
77   -73.897746  40.677548    0      False       True   working    NY   
79   -73.947781  40.632336    0      False       True   working    NY   
81   -74.027731  40.622159    0       True       True    broken    NY   
85   -73.890678  40.676235    0      False       True   working    NY   
92   -73.928650  40.632080    0      False       True   working    NY   
93   -73.982658  40.612892    0      False       True   working    NY   
94   -73.879116  40.679871    0      False       True   working    NY   
96   -73.889664  40.657864    0      False       True   working    NY   
97   -73.915535  40.633762    0      False       True   working    NY   
98   -73.901260  40.644558    0      False       True   working    NY   
103  -73.998260  40.604786    0      False       True   working    NY   
107  -73.928970  40.615036    0      False       True   working    NY   
108  -73.989006  40.599335    0      False       True   working    NY   
116  -73.860352  40.670147    0      False       True   working    NY   
117  -73.889877  40.633153    0      False       True   working    NY   
118  -73.961494  40.598076    0      False       True   working    NY   
125  -73.941605  40.597485    0      False       True   working    NY   
134  -73.975105  40.579582    0      False       True   working    NY   
135  -73.981818  40.577218    0       True       True    broken    NY   
2681 -81.734882  41.426374    0       True       True    broken    OH   
2688 -81.758967  41.419002    0      False       True   working    OH   
3726 -84.248451  42.099999    0      False      False  inactive    MI   

          city                   street country             last_checked  
4     Brooklyn              82 Court St     USA  Checked 142 minutes ago  
10    Brooklyn   395 Flatbush Ave Exten     USA  Checked 142 minutes ago  
19    Brooklyn             267 Broadway     USA  Checked 142 minutes ago  
23    Brooklyn         357 Hamilton Ave     USA  Checked 142 minutes ago  
27    Brooklyn        904 Manhattan Ave     USA  Checked 142 minutes ago  
28    Brooklyn         840 Atlantic Ave     USA  Checked 

Untuk Query, selain spesifik untuk minta 1 to 1 match, bisa juga ada kondisi seperti:

- .isin, salah satu dari beberapa pilihan
- .str.contains, ada substring tertentu
- operator matematik untuk data numerik, /> /< />= /<= /!=


In [86]:
df_mcd['city'].unique()

array(['New York', 'Brooklyn', 'Jersey City', ..., 'Kiefersfelden',
       'Herleshausen', 'Haren'], dtype=object)

In [87]:
# boleh dua city ini
df_mcd[df_mcd['city'].isin(['Brooklyn', 'New York'])]

lat        lon  alt  is_broken  is_active       dot state  \
0    -73.988281  40.718830    0      False       True   working    NY   
1    -74.005090  40.728794    0      False       True   working    NY   
2    -73.993408  40.729197    0      False       True   working    NY   
3    -73.985855  40.726555    0      False       True   working    NY   
4    -73.991692  40.691383    0       True       True    broken    NY   
...         ...        ...  ...        ...        ...       ...   ...   
174  -73.921301  40.867396    0      False       True   working    NY   
190  -73.843681  40.840374    0      False       True   working    NY   
2681 -81.734882  41.426374    0       True       True    broken    OH   
2688 -81.758967  41.419002    0      False       True   working    OH   
3726 -84.248451  42.099999    0      False      False  inactive    MI   

          city                 street country             last_checked  
0     New York        114 Delancey St     USA  Checked 142 minutes ago  
1     New York          208 Varick St     USA  Checked 142 minutes ago  
2     New York           724 Broadway     USA  Checked 142 minutes ago  
3     New York            102 1st Ave     USA  Checked 142 minutes ago  
4     Brooklyn            82 Court St     USA  Checked 142 minutes ago  
...        ...                    ...     ...                      ...  
174   New York  608 West 207th Street     USA  Checked 144 minutes ago  
190   New York  16 Westchester Square     USA  Checked 145 minutes ago  
2681  Brooklyn          4738 Ridge Rd     USA  Checked 168 minutes ago  
2688  Brooklyn     10700 Brookpark Rd     USA  Checked 169 minutes ago  
3726  Brooklyn          306 S Main St     USA   Checked 13 minutes ago  

[90 rows x 11 columns]

In [102]:
# namanya nya dimulai dengan huruf H
kondisi5 = df_superstore['Customer Name'].str.startswith('H')
df_superstore[kondisi5]

Row ID         Order ID Order Date Dispatch Date   Delivery Mode  \
178      179  ES-2020-5980468 2020-08-30    2020-09-04  Standard Class   
184      185  ES-2020-4698340 2020-09-01    2020-09-02     First Class   
524      525  ES-2021-4548757 2021-12-16    2021-12-23  Standard Class   
603      604  ES-2022-5949760 2022-08-29    2022-09-02  Standard Class   
604      605  ES-2022-5949760 2022-08-29    2022-09-02  Standard Class   
...      ...              ...        ...           ...             ...   
9726    9727  ES-2022-3994900 2022-08-29    2022-09-01    Second Class   
9727    9728  ES-2022-3994900 2022-08-29    2022-09-01    Second Class   
9878    5215  ES-2021-4513317 2021-05-24    2021-05-26     First Class   
9879    5216  ES-2021-4513317 2021-05-24    2021-05-26     First Class   
9880    5217  ES-2021-4513317 2021-05-24    2021-05-26     First Class   

     Customer ID      Customer Name    Segment       City  \
178     HA-14920     Helen Andreada   Consumer        Ulm   
184     HF-14995    Herbert Flentye   Consumer      Paris   
524     HK-14890   Heather Kirkland  Corporate   Stafford   
603     HA-14905      Helen Abelman   Consumer  Barcelona   
604     HA-14905      Helen Abelman   Consumer  Barcelona   
...          ...                ...        ...        ...   
9726    HM-14980  Henry MacAllister   Consumer       Rome   
9727    HM-14980  Henry MacAllister   Consumer       Rome   
9878    HG-15025      Hunter Glantz   Consumer  Wuppertal   
9879    HG-15025      Hunter Glantz   Consumer  Wuppertal   
9880    HG-15025      Hunter Glantz   Consumer  Wuppertal   

              State/Province  Country/Region   Region       Product ID  \
178        Baden-Württemberg         Germany  Central  FUR-CH-10002053   
184            Ile-de-France          France  Central  TEC-CO-10002269   
524                  England  United Kingdom    North  OFF-PA-10000850   
603                Catalonia           Spain    South  TEC-AC-10003666   
604                Catalonia           Spain    South  FUR-BO-10004119   
...                      ...             ...      ...              ...   
9726                   Lazio           Italy    South  OFF-BI-10002570   
9727                   Lazio           Italy    South  OFF-AR-10000475   
9878  North Rhine-Westphalia         Germany  Central  OFF-AR-10001002   
9879  North Rhine-Westphalia         Germany  Central  OFF-AR-10003457   
9880  North Rhine-Westphalia         Germany  Central  OFF-AR-10001529   

             Category Sub-Category  \
178         Furniture       Chairs   
184        Technology      Copiers   
524   Office Supplies        Paper   
603        Technology  Accessories   
604         Furniture    Bookcases   
...               ...          ...   
9726  Office Supplies      Binders   
9727  Office Supplies          Art   
9878  Office Supplies          Art   
9879  Office Supplies          Art   
9880  Office Supplies          Art   

                                     Product Name    Sales  Quantity  \
178                  Novimex Chairmat, Set of Two  263.655         5   
184                            Brother Ink, Color  632.145         5   
524           SanDisk Parchment Paper, Multicolor   54.360         3   
603                     Memorex Router, Bluetooth  492.420         2   
604          Ikea Floating Shelf Set, Traditional  851.700         5   
...                                           ...      ...       ...   
9726                    Acco Binder Covers, Clear   25.560         2   
9727                         Sanford Canvas, Blue  101.040         2   
9878                  BIC Sketch Pad, Fluorescent  349.440         7   
9879              Sanford Sketch Pad, Water Color   97.320         2   
9880  Binney & Smith Pencil Sharpener, Easy-Erase  167.580         6   

      Discount   Profit  
178       0.10   17.505  
184       0.15  252.795  
524       0.00   10.800  
603       0.00  187.080  
604       0.00   51.000  
...        ...   

In [103]:
df_filtered = df_superstore[kondisi5]
df_filtered['Customer Name'].value_counts()

Customer Name
Harry Marie          31
Harry Greene         22
Henia Zydlo          20
Henry MacAllister    19
Heather Jas          19
Heather Kirkland     18
Herbert Flentye      18
Hunter Glantz        14
Helen Wasserman      13
Hilary Holden        13
Harold Dahlen        12
Harold Engle         12
Hallie Redmond       11
Henry Goldwyn         9
Helen Andreada        9
Harold Ryan           8
Helen Abelman         7
Harold Pawlan         6
Hunter Lopez          5
Name: count, dtype: int64

In [105]:
df_filtered['Customer Name'].str.slice(0, 1)

178     H
184     H
524     H
603     H
604     H
       ..
9726    H
9727    H
9878    H
9879    H
9880    H
Name: Customer Name, Length: 266, dtype: object

In [107]:
df_filtered['first_letter'] = df_filtered['Customer Name'].str.slice(0, 1)
df_filtered

C:\Users\user\AppData\Local\Temp\ipykernel_29320\2863482511.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_filtered['first_letter'] = df_filtered['Customer Name'].str.slice(0, 1)


Row ID         Order ID Order Date Dispatch Date   Delivery Mode  \
178      179  ES-2020-5980468 2020-08-30    2020-09-04  Standard Class   
184      185  ES-2020-4698340 2020-09-01    2020-09-02     First Class   
524      525  ES-2021-4548757 2021-12-16    2021-12-23  Standard Class   
603      604  ES-2022-5949760 2022-08-29    2022-09-02  Standard Class   
604      605  ES-2022-5949760 2022-08-29    2022-09-02  Standard Class   
...      ...              ...        ...           ...             ...   
9726    9727  ES-2022-3994900 2022-08-29    2022-09-01    Second Class   
9727    9728  ES-2022-3994900 2022-08-29    2022-09-01    Second Class   
9878    5215  ES-2021-4513317 2021-05-24    2021-05-26     First Class   
9879    5216  ES-2021-4513317 2021-05-24    2021-05-26     First Class   
9880    5217  ES-2021-4513317 2021-05-24    2021-05-26     First Class   

     Customer ID      Customer Name    Segment       City  \
178     HA-14920     Helen Andreada   Consumer        Ulm   
184     HF-14995    Herbert Flentye   Consumer      Paris   
524     HK-14890   Heather Kirkland  Corporate   Stafford   
603     HA-14905      Helen Abelman   Consumer  Barcelona   
604     HA-14905      Helen Abelman   Consumer  Barcelona   
...          ...                ...        ...        ...   
9726    HM-14980  Henry MacAllister   Consumer       Rome   
9727    HM-14980  Henry MacAllister   Consumer       Rome   
9878    HG-15025      Hunter Glantz   Consumer  Wuppertal   
9879    HG-15025      Hunter Glantz   Consumer  Wuppertal   
9880    HG-15025      Hunter Glantz   Consumer  Wuppertal   

              State/Province  ...   Region       Product ID         Category  \
178        Baden-Württemberg  ...  Central  FUR-CH-10002053        Furniture   
184            Ile-de-France  ...  Central  TEC-CO-10002269       Technology   
524                  England  ...    North  OFF-PA-10000850  Office Supplies   
603                Catalonia  ...    South  TEC-AC-10003666       Technology   
604                Catalonia  ...    South  FUR-BO-10004119        Furniture   
...                      ...  ...      ...              ...              ...   
9726                   Lazio  ...    South  OFF-BI-10002570  Office Supplies   
9727                   Lazio  ...    South  OFF-AR-10000475  Office Supplies   
9878  North Rhine-Westphalia  ...  Central  OFF-AR-10001002  Office Supplies   
9879  North Rhine-Westphalia  ...  Central  OFF-AR-10003457  Office Supplies   
9880  North Rhine-Westphalia  ...  Central  OFF-AR-10001529  Office Supplies   

     Sub-Category                                 Product Name    Sales  \
178        Chairs                 Novimex Chairmat, Set of Two  263.655   
184       Copiers                           Brother Ink, Color  632.145   
524         Paper          SanDisk Parchment Paper, Multicolor   54.360   
603   Accessories                    Memorex Router, Bluetooth  492.420   
604     Bookcases         Ikea Floating Shelf Set, Traditional  851.700   
...           ...                                          ...      ...   
9726      Binders                    Acco Binder Covers, Clear   25.560   
9727          Art                         Sanford Canvas, Blue  101.040   
9878          Art                  BIC Sketch Pad, Fluorescent  349.440   
9879          Art              Sanford Sketch Pad, Water Color   97.320   
9880          Art  Binney & Smith Pencil Sharpener, Easy-Erase  167.580   

      Quantity  Discount   Profit  first_letter  
178          5      0.10   17.505             H  
184          5      0.15  252.795             H  
524          3      0.00   10.800             H  
603          2      0.00  187.080             H  
604          5      0.00   51.000             H  
...        ...       ...      ...           ...  
9726         2      0.00   10.980             H  
9727         2      0.00   29.280             H  
9878         7      0.00  115.290             H  
9879         2      0.

# Latihan

kita coba gunakan dataset penjualan kopi dari https://www.kaggle.com/datasets/ihelon/coffee-sales

1. convert 1 kolom harga dari currency ukraina jadi rupiah
2. cek ada berapa jenis coffee name, lalu masing2 terjual berapa kali
3. lihat rata2 dan jumlah harga penjualan per jenis kopi
